packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -2727,37 +2734,37 @@ export class SnapController extends BaseController<
       "'permissionName' must be either a string or null.",
     );
 
-    if (permissionName) {
-      const hasPermission = this.messagingSystem.call(
-        'PermissionController:hasPermission',
-        snapId,
-        permissionName,
-      );
+    const permissions = this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      snapId,
+    );
 
-      if (!hasPermission) {
-        throw new Error(
-          `Snap "${snapId}" is not permitted to use "${permissionName}".`,
-        );
-      }
+    // If permissionName is null, the handler does not require a permission.
+    if (
+      permissionName !== null &&
+      (!permissions || !hasProperty(permissions, permissionName))
+    ) {
+      throw new Error(
+        `Snap "${snapId}" is not permitted to use "${permissionName}".`,
+      );
     }
 
+    const handlerPermissions = permissionName
+      ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        permissions![permissionName]
+      : undefined;
+
     if (
       permissionName === SnapEndowments.Rpc ||
       permissionName === SnapEndowments.Keyring
     ) {
+      assert(handlerPermissions);
```

Comment:
Should this assertion be above the `if`, given that it's also been used with ` const timeout = this.#getExecutionTimeout(handlerPermissions);` at the end?

Reply:
It should not, `handlerPermissions` is only required here and is fine to be undefined in `#getExecutionTimeout`.